rpk: show decommissioning progress

daisukebe opened the pull request:
<!--

See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md##pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED.

Describe, in plain language, the motivation behind the change (bug fix,
feature, improvement) in this PR and how the included commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.

  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.

  Backport of PR #PR-NUMBER

-->

Introduce a new command `rpk redpanda admin brokers decommission-status <node-id>`. This relies on the feature added by https://github.com/redpanda-data/redpanda/pull/8167.
Fixes https://github.com/redpanda-data/redpanda/issues/8268

Example outputs:

If the node is not decommissioning,
```
$ rpk redpanda admin brokers decommission-status 4
Node 4 is not decommissioning
```
If the node is in decommissioning, it prints below by default
```
$ rpk redpanda admin brokers decommission-status 4
NAMESPACE-TOPIC              PARTITION  MOVING-TO  COMPLETION-%  PARTITION-SIZE
kafka/test                   0          3          9             1699470920
kafka/test                   4          3          0             1614258779
kafka/test2                  3          3          3             2722706514
kafka/test2                  4          3          4             2945518089
kafka_internal/id_allocator  0          3          0             3562
```

With --detailed/-d, it prints granular outputs below.
```
$ rpk redpanda admin brokers decommission-status 4 -d
NAMESPACE-TOPIC  PARTITION  MOVING-TO  COMPLETION-%  PARTITION-SIZE  BYTES-MOVED  BYTES-REMAINING
kafka/test       0          3          13            1731773243      228114727    1503658516
kafka/test       4          3          1             1645952961      18752660     1627200301
kafka/test2      3          3          5             2752632301      140975805    2611656496
kafka/test2      4          3          6             2975443783      181581219    2793862564
```

Once completed, it reports
```
$ rpk redpanda admin brokers decommission-status 4
Node 4 is decommissioned successfully.
```
## Backports Required

<!--

Checking at least one of the checkboxes is REQUIRED if this PR is not a backport.

-->

- [ ] none - not a bug fix
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v22.3.x
- [ ] v22.2.x
- [ ] v22.1.x

## UX Changes

<!--

Content in this section is OPTIONAL.

Describe, in plain language, how this PR affects an end-user. Explain
topic flags, configuration flags, command line flags, deprecation
policies, etc. that are added or modified. Don't ship user breaking
changes. Ask the @redpanda-data/product team if you need help with user
visible changes.

-->

A new command is added to make the decommissioning visibility better

## Release Notes

  ### Features

  * rpk: Introduce a new command `rpk redpanda admin brokers decommission-status <node-id>`

bharathv commented:
this is very nice, just a drive by comment, should we pretty print the byte columns? We already have % completion, the operator may not care about raw byte values.

twmb commented on src/go/rpk/pkg/cli/cmd/redpanda/admin/brokers/brokers.go:
```diff
@@ -161,3 +162,39 @@ the cluster leader handles the request.
+			dbs, err := cl.DecommissionBrokerStatus(cmd.Context(), broker)
+			out.MaybeDie(err, "unable to request brokers: %v", err)
+
+			headers := []string{"Namespace-Topic", "Partition", "Moving-to", "Parttion-size", "Bytes-moved", "Bytes-remaining", "Completion (%)"}
```
typo `parttion`

also I'd change `completion (%)` to `completion-%` -- where possible, I try to format the output for awk

daisukebe replied:
Great point.

twmb commented on src/go/rpk/pkg/cli/cmd/redpanda/admin/brokers/brokers.go:
```diff
@@ -161,3 +162,39 @@ the cluster leader handles the request.
+				if p.PartitionSize > 0 {
+					completion = p.BytesMoved * 100 / p.PartitionSize
+				}
+				ret := []interface{}{nt, p.Partition, p.MovingTo.NodeID, p.PartitionSize, p.BytesMoved, p.BytesLeftToMove, completion}
```
Can this change to:
```
return struct {
    // types
}{
    // fields
}
```
with types & fields each on a dedicated line? And then below, use `PrintStructFields`?

This way, we ensure that any type changes in api_broker cause compile time failures here -- this file can guard itself against API changes in the other file.

(I introduced PrintStructFields after I accidentally caused a type change from a number to a pointer)

daisukebe replied:
Wasn't aware of it, thank you!

daisukebe commented:
I have addressed @twmb 's comments in the latest commit, e255380 
Besides, I've added some case handling below. Does this make sense?

If the node is _not_ in `decommissioning`, it returns
```
$ rpk redpanda admin brokers decommission-status 2
Node 2 is not decommissioning
```
If the node is decommissioned, it returns
```
$ rpk redpanda admin brokers decommission-status 2
The node is decommissioned. 0 replicas left
```

@bharathv , thanks for reviewing. I thought the same originally but it turned out that it's better to print the raw figures in a case users want to see the progress in details. Given the `completion-%` is just two digits, it may not change for a while in a large cluster. In that scenario, checking the raw bytes will be helpful rather than letting users invoke the raw admin API. Wdyt?

daisukebe commented:
It may also make sense to hide the raw byte values by default and let them reveal with `-v` option.

Additionally I will split the function to an individual file upon https://github.com/redpanda-data/redpanda/pull/8371 gets committed

twmb commented on src/go/rpk/pkg/cli/cmd/redpanda/admin/brokers/brokers.go:
```diff
@@ -161,3 +163,64 @@ the cluster leader handles the request.
+
+			dbs, err := cl.DecommissionBrokerStatus(cmd.Context(), broker)
+			var he *admin.HTTPResponseError
+			if errors.As(err, &he) {
```
If the error is not HTTPResponseError, we need to exit with the untyped error

twmb commented on src/go/rpk/pkg/cli/cmd/redpanda/admin/brokers/brokers.go:
```diff
@@ -161,3 +163,64 @@ the cluster leader handles the request.
+				body, bodyErr := he.DecodeGenericErrorBody()
+				if bodyErr == nil {
+					out.Die("%s", body.Message)
+				}
```
What if bodyErr is non-nil?

twmb commented on src/go/rpk/pkg/cli/cmd/redpanda/admin/brokers/brokers.go:
```diff
@@ -161,3 +163,64 @@ the cluster leader handles the request.
+				}
+			}
+			if dbs.Finished {
+				out.Die("The node is decommissioned. %d replicas left", dbs.ReplicasLeft)
```
Why die here? die exits with 1

also I don't understand the second sentence / what's it mean?

twmb replied:
Actually I think I understand it -- is there any instance where there will be more than 0 replicas left?

daisukebe replied:
> is there any instance where there will be more than 0 replicas left?

In theory, it should be always 0 as long as it's `finished`. I wanted to utilize the raw json output and make it easier to catch a bug in a case non-zero value is returned whereas decommission is marked `finished`.

A raw output when decommissioning is finished.
```
{
  "finished": true,
  "replicas_left": 0
}
```
Do you think if's too much?

daisukebe replied:
I'd lean toward the following way:
If `finished` is true and
* if `replicas_left == 0`, print `The node is decommissioned successfully`
* else, print `The node is decommissioned but xx replicas left`

Wdyt?

twmb replied:
What should a user do if xx replicas are left?

daisukebe replied:
That's so far unknown, tbh. It's most likely a bug in Redpanda if that happens.

daisukebe replied:
Should we encourage the users to report to us in GH if they hit this?
```
The node is decommissioned but %d replicas left. There is likely an issue inside Redpanda. Please describe how you encountered this at https://github.com/redpanda-data/redpanda/issues/new?assignees=&labels=kind%2Fbug&template=01_bug_report.md
```

daisukebe replied:
@mmaslankaprv , what do you think is a better advise in a case replicas_left is non-zero whereas decommission is marked finished? Shouldn't we consider about that kind of abnormal case in general?

daisukebe replied:
Nvm, I've removed the check in a case `replicas_left` is non-zero since it's must be a quite rare case.

twmb replied:
We have one other area inside rpk where we print "if you run into this, please file a bug" -- so I'm open to that, too, fwiw.

daisukebe replied:
Yea, I've found it so that come with the idea. I've reverted the check for better supportability. Thanks!

daisukebe commented:
New changes:

* Split the command function to a new dedicated file
* Sorted by topic using the `types` package
* Introduced `--detailed` option. 
 
By default it prints these
```
$ rpk redpanda admin brokers decommission-status 4
NAMESPACE-TOPIC              PARTITION  MOVING-TO  COMPLETION-%  PARTITION-SIZE
kafka/test                   0          3          9             1699470920
kafka/test                   4          3          0             1614258779
kafka/test2                  3          3          3             2722706514
kafka/test2                  4          3          4             2945518089
kafka_internal/id_allocator  0          3          0             3562
```
with `--detailed/-d`, it prints granular outputs below.

```
$ rpk redpanda admin brokers decommission-status 4 -d
NAMESPACE-TOPIC  PARTITION  MOVING-TO  COMPLETION-%  PARTITION-SIZE  BYTES-MOVED  BYTES-REMAINING
kafka/test       0          3          13            1731773243      228114727    1503658516
kafka/test       4          3          1             1645952961      18752660     1627200301
kafka/test2      3          3          5             2752632301      140975805    2611656496
kafka/test2      4          3          6             2975443783      181581219    2793862564
```

twmb commented on src/go/rpk/pkg/cli/cmd/redpanda/admin/brokers/decommission-status.go:
```diff
@@ -0,0 +1,106 @@
+			out.MaybeDie(err, "unable to request brokers: %v", err)
+
+			if dbs.Finished {
+				out.Exit("The node is decommissioned successfully")
```
Can this end in a period?

twmb reviewed (dismissed):
one nit, lgtm though.

twmb commented on src/go/rpk/pkg/cli/cmd/redpanda/admin/brokers/decommission-status.go:
```diff
@@ -0,0 +1,110 @@
+				if dbs.ReplicasLeft == 0 {
+					out.Exit("Note %d is decommissioned successfully.", broker)
+				} else {
+					out.Exit("Node %d is decommissioned but %d replicas left. So that there is likely an issue inside Redpanda. Please describe how you encountered this at https://github.com/redpanda-data/redpanda/issues/new?assignees=&labels=kind%2Fbug&template=01_bug_report.md", broker, dbs.ReplicasLeft)
```
The wording we use in config/params.go is:
```go
"rpk bug, please describe how you encountered this at https://github.com/redpanda-data/redpanda/issues/new?assignees=&labels=kind%2Fbug&template=01_bug_report.md"
```

wdyt about:
```go
out.Exit("Node %d is decommissioned but there are %d replicas left, which may be an issue inside Redpanda. Please describe how you encountered this at https://github.com/redpanda-data/redpanda/issues/new?assignees=&labels=kind%2Fbug&template=01_bug_report.md")
```

daisukebe replied:
Makes sense. Thank you.

daisukebe commented:
Build failure looks unrelated.

daisukebe commented:
@piyushredpanda @r-vasquez can we merge this and release as a part of 23.1 if it’s not too late?

piyushredpanda commented:
@daisukebe : Of course it should be in v23.1, but also I thought this should go along with decommission progress endpoint being backported to v22.3.x, no?

daisukebe commented:
Thanks @piyushredpanda . Regarding backporint go v22.3.x, I agreed but there are two blockers to make this work, https://github.com/redpanda-data/redpanda/pull/8265 and https://github.com/redpanda-data/redpanda/pull/8347

bharathv commented:
@daisukebe I merged Michal's backport, fyi.

daisukebe commented:
@bharathv , cool, thanks! Are you able to merge this as well, by any chance, in a case the build failure is irrelevant? Rogger and I are not able to merge in GH (grayed out)

bharathv commented:
/ci-repeat 1

bharathv commented:
>  Are you able to merge this as well, by any chance, in a case the build failure is irrelevant

Failures seem unrelated but I kicked off another run just in case. @piyushredpanda can merge if it runs into any unrelated issues.

piyushredpanda commented:
@daisukebe / @bharathv : Hmm, I see a failure with, which does not seem like a known issue?

```
Module: rptest.tests.random_node_operations_test
Class:  RandomNodeOperationsTest
Method: test_node_operations
Arguments:
{
  "enable_failures": true
```

r-vasquez commented:
Seems unrelated, the test uses some known commands (either calling the admin API directly or via rpk) but this PR only introduces a new command without changing the behavior of an already existing one.

Let me run this locally to see if I can reproduce, if not, I'll create the issue and link it to this PR @piyushredpanda

r-vasquez commented:
/ci-repeat 1
skip-units
dt-repeat=10
tests/rptest/tests/random_nodes_operations_test.py::RandomNodeOperationsTest.test_node_operations

bharathv commented:
> Hmm, I see a failure with, which does not seem like a known issue?

That looks like #8589

r-vasquez commented:
My ci-repeat call failed, sorry; I had to cancel and[ force push](https://github.com/redpanda-data/redpanda/compare/340023e8f18e73b18610eb35eeb90c3479c69bf8..e2523b679c7e237c0eef8cdacbe4b85408495380) with no changes to create another build. 

The original error was unrelated (the one that Bharat posted)

bharathv commented:
/backport v22.3.x

vbotbuildovich commented:
The pull request is not merged yet. Cancelling backport...

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/4088960996).

daisukebe commented:
/backport v22.3.x

piyushredpanda commented:
@daisukebe : @bharathv created https://github.com/redpanda-data/redpanda/pull/8625 as a manual backport to speed things up already.

daisukebe commented:
Oops. I wasn’t aware of that. Thank you!
